src/instrumentation/podmutator.go

```diff
@@ -35,150 +45,250 @@ var (
+			pm.logger.Info("too many New Relic Instrumentation instances for this Pod.  only 1 allowed")
+			return pod, nil
+		} else {
+			return pod, err
+		}
 	}
-	insts.NodeJS = inst
+	mutatedPod := pm.sdkInjector.Inject(ctx, instrumentations, ns, pod)
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectPython); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	if licenseKeySecret, err := pm.getLicenseKeySecretFn(instCandidates); err != nil {
+		logger.Error(err, "failed to identify the correct secret.  all matching instrumentation's must use the same secret")
+		return pod, nil
+	} else {
+		err = pm.secretReplicator.ReplicateSecret(ctx, ns, pod, pm.operatorNamespace, licenseKeySecret)
+		if err != nil {
+			logger.Error(err, "failed to replicate secret")
+			return pod, nil
+		}
 	}
-	insts.Python = inst
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectDotNet); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+	return mutatedPod, nil
+}
+
+// GetLanguageInstrumentations is used to collect all instrumentations and validate that only a single insturmentation
+// exists for each language, and return them together
+func GetLanguageInstrumentations(instCandidates []*v1alpha2.Instrumentation) ([]*v1alpha2.Instrumentation, error) {
+	var instrumentations []*v1alpha2.Instrumentation
+	languages := map[string]*v1alpha2.Instrumentation{}
+	for _, candidate := range instCandidates {
+		if currentInst, ok := languages[candidate.Spec.Agent.Language]; ok {
+			if currentInst.Spec.Agent.Image != candidate.Spec.Agent.Image ||
+				!reflect.DeepEqual(currentInst.Spec.Agent.Env, candidate.Spec.Agent.Env) ||
+				!reflect.DeepEqual(currentInst.Spec.Agent.VolumeSizeLimit, candidate.Spec.Agent.VolumeSizeLimit) ||
+				!reflect.DeepEqual(currentInst.Spec.Agent.Resources, candidate.Spec.Agent.Resources) {
+				return nil, errMultipleInstancesPossible
+			}
+		} else {
+			languages[candidate.Spec.Agent.Language] = candidate
+			instrumentations = append(instrumentations, candidate)
+		}
 	}
-	insts.DotNet = inst
+	return instrumentations, nil
+}
 
-	if inst, err = pm.getInstrumentationInstance(ctx, ns, pod, annotationInjectPhp); err != nil {
-		// we still allow the pod to be created, but we log a message to the operator's logs
-		logger.Error(err, "failed to select a New Relic Instrumentation instance for this pod")
-		return pod, err
+// InstrumentationLocator is used to find instrumentations
+type InstrumentationLocator interface {
+	GetInstrumentations(ctx context.Context, ns corev1.Namespace, pod corev1.Pod) ([]*v1alpha2.Instrumentation, error)
+}
```

Comment:
I don't think we need this interface. All you need is the client for the function `GetInstrumentations`. Why can't we just define the function `GetInstrumentations` on the `instPodMutator`. Is this  a workaround for testing?

Reply:
The is specifically for testing